### DISCUSSION TITLE
Relax ```new()``` and ```required``` mutual exclusion

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#7947] [Language Ideas]
Today when a type has a ```required``` property, the type cannot fulfill the ```new()``` constraint.

This requirements should be removed when all the following are true:
* Additional constraints specify all required properties for the type.

Example:
```
public record LegacyRecord {
  public required long Id {get; init; }
}

pubic record D1 : LegacyRecord {
  public string? Name {get; init;}
}

pubic  record D2 : LegacyRecord {
  public required string? Name {get; init;}
}

public static void Test<T> where T : LegacyRecord, new() {
  var ret = new T(){
    Id = 0,
  };
  return ret;
}

Test<LegacyRecord>(); //this should work.
Test<D1>(); //this also should work.
Test<D2>(); //this should not work because D2 introduces required members that are not covered by a constraint..
``` 

Thanks!

**Comment by TahirAhmadov** [comment]
> `Test<LegacyRecord>(); //this should work.`

This cannot work because the compiler does not know that `Test` sets the required property `Id` in its body - imagine that it lives in a different DLL and C# does not interpret the contents, and it can be an interface method so there is no body at all.
For this to work, there would need to be something in the signature - the `new()` constraint would get additional metadata:
```cs
public static void Test<T> where T : LegacyRecord, new() sets_required_properties // new keyword here
{
  var ret = new T() { Id = 0, }; // no error
  var x = new T(); // error - Id not set
  return ret;
}
```
The imaginary new `sets_required_properties` keyword specifies that 1) a base class constraint is also required, and 2) all of the required properties of said base class are set in the generic method.

**Reply by TonyValenti** [reply]
I'm not sure I understand. 

By combining new() and a base class constraint, I should be able to call my method with any T that does not add required members to the base class and I should be able to know that all required properties are set because it is the same list as the base class.

**Reply by TahirAhmadov** [reply]
You're forgetting about constructors on child classes which have `[SetsRequiredMembers]` attribute.
```cs
pubic  record D3 : LegacyRecord {
  [SetsRequiredMembers]
  public D3() { ... }
}
public static void Test2<T> where T : LegacyRecord, new()
{
  var ret = new T() { Id = 0, }; // no error
  var x = new T(); // no error as well
  return ret;
}
Test2<D3>(); // this works
```

**Reply by TonyValenti** [reply]
Oh!  I see. But that would not prevent the scenario I described. It would just result in double initialization for some properties in some cases.

**Reply by 333fred** [reply]
It's not that it cannot work. It's that it would require a much more complex analysis, one we weren't sure we were ready to define. We've also not seen much ask for it since `required` was introduced; I went looking to see if this was a duplicate and couldn't find any direct duplicate discussions.

**Reply by TonyValenti** [reply]
Thanks @333fred !
Here are details on my scenario and what triggered this need:
I have classes that have base-class properties that are always set at construction.  I wanted to make those properties required so that I could prevent any bugs from cropping up where one of the properties might accidentally not be set. I could not.

**Reply by qrli** [reply]
This scenario is similar to `new()` constraint with non-empty parameter list.

**Reply by TonyValenti** [reply]
@qrli Excellent parallel!  You're exactly right!
